### PR TITLE
SNOW-675138: Add telemetry to queries that use sql simplifier

### PR DESCRIPTION
**Pull request by sfc-gh-yixie**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-675138

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.
     Instrument 1) when the sql simplifier is enabled 2) when a df action is invoked

### CODE REVIEW

File: src/snowflake/snowpark/_internal/telemetry.py [hunk 1]
```diff
@@ -359,3 +363,15 @@ def send_copy_pattern_telemetry(self):
         self.send_function_usage_telemetry(
             "copy_pattern", TelemetryField.FUNC_CAT_COPY.value
         )
+
+    def send_sql_simplifier_telemetry(self, session_id, sql_simplifier_enabled):
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/telemetry.py:367]:
```suggestion
    def send_sql_simplifier_telemetry(self, session_id: str, sql_simplifier_enabled: bool) -> None:
```

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -319,6 +317,19 @@ def close(self) -> None:
...
 
+    @property
+    def sql_simplifier_enabled(self) -> bool:
+        return self._sql_simplifier_enabled
+
+    @sql_simplifier_enabled.setter
+    def sql_simplifier_enabled(self, value: bool) -> None:
+        """Set to ``True`` to use the SQL simplifier.
+        The generated SQLs from ``DataFrame`` transformations would have fewer layers of nested queries if the SQL simplifier is enabled."""
+        self._conn._telemetry_client.send_sql_simplifier_telemetry(
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/session.py:328]: Should we also check `telemetry_enabled` here? If `telemetry_enabled` is False, we shouldn't send the telemetry.

**Review by sfc-gh-aalam — APPROVED** [review]
LGTM

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/session.py:328]: This method itself checks the whether telemetry is enabled.

**Review by sfc-gh-jdu — APPROVED** [review]

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/_internal/telemetry.py [hunk 3]
```diff
@@ -137,6 +140,7 @@ def wrap(*args, **kwargs):
             *plan.api_calls,
             {TelemetryField.NAME.value: f"DataFrame.{func.__name__}"},
         ]
+        api_calls[0]["sql_simplifier_enabled"] = args[0]._session.sql_simplifier_enabled
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/telemetry.py:143]: minor: let's add a comment why we insert it at the 0th item?

File: src/snowflake/snowpark/_internal/telemetry.py [hunk 4]
```diff
@@ -359,3 +363,15 @@ def send_copy_pattern_telemetry(self):
         self.send_function_usage_telemetry(
             "copy_pattern", TelemetryField.FUNC_CAT_COPY.value
         )
+
+    def send_sql_simplifier_telemetry(self, session_id: str, sql_simplifier_enabled: bool) -> None:
+        message = {
+            **self._create_basic_telemetry_data(
+                TelemetryField.TYPE_FUNCTION_USAGE.value
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/_internal/telemetry.py:370]: I think we shall introduce a new type of telemetry: TYPE_SIMPLIFIER_UAGE, the function usage is for telemetry of api calls.

**Pull request merged**